File: airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/java/io/airbyte/cdk/integrations/destination/s3/csv/RootLevelFlatteningSheetGenerator.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.integrations.destination.s3.csv;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.collect.Lists;
import io.airbyte.cdk.integrations.base.JavaBaseConstants;
import io.airbyte.cdk.protocol.deser.PartialJsonDeserializer;
import io.airbyte.cdk.protocol.deser.StringIterator;
import io.airbyte.commons.json.Jsons;
import io.airbyte.commons.util.MoreIterators;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public class RootLevelFlatteningSheetGenerator extends BaseSheetGenerator implements CsvSheetGenerator {

  /**
   * Keep a header list to iterate the input json object with a defined order.
   */
  private final List<String> recordHeaders;

  public RootLevelFlatteningSheetGenerator(final JsonNode jsonSchema) {
    this.recordHeaders = MoreIterators.toList(jsonSchema.get("properties").fieldNames())
        .stream().sorted().collect(Collectors.toList());
  }

  @Override
  public List<String> getHeaderRow() {
    final List<String> headers = Lists.newArrayList(JavaBaseConstants.COLUMN_NAME_AB_ID,
        JavaBaseConstants.COLUMN_NAME_EMITTED_AT);
    headers.addAll(recordHeaders);
    return headers;
  }

  /**
   * With root level flattening, the record columns are the first level fields of the json.
   */
  @Override
  List<String> getRecordColumns(final String serializedJson) {
    final Map<String, String> fields = PartialJsonDeserializer.parseObject(new StringIterator(serializedJson), recordHeaders);

    final List<String> output = new ArrayList<>();
    for (final String field : recordHeaders) {
      final String serializedValue = fields.getOrDefault(field, "");
      if (serializedValue.startsWith("\"")) {
        // Unwrap string values, i.e. we want to remove the double quotes,
        // handle escape characters, etc.
        // TODO it would be cool to do this within the deserializer, i.e. 1-pass processing
        output.add(Jsons.deserialize(serializedValue).asText());
      } else if ("null".equals(serializedValue)) {
        // Write empty string instead of null
        output.add("");
      } else {
        // For other values (numbers, booleans, arrays, objects) just write the raw serialized value
        output.add(serializedValue);
      }
    }
    return output;
  }

}

File: airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/csv/RootLevelFlatteningSheetGenerator.kt
/*
 * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.integrations.destination.s3.csv

import com.fasterxml.jackson.databind.JsonNode
import io.airbyte.cdk.integrations.base.JavaBaseConstants
import io.airbyte.commons.json.Jsons
import io.airbyte.commons.util.MoreIterators
import java.util.LinkedList

class RootLevelFlatteningSheetGenerator(
    jsonSchema: JsonNode,
    private val useV2FieldNames: Boolean = false
) : BaseSheetGenerator(useV2FieldNames), CsvSheetGenerator {
    /** Keep a header list to iterate the input json object with a defined order. */
    private val recordHeaders: List<String> =
        MoreIterators.toList(
                jsonSchema["properties"].fieldNames(),
            )
            .sorted()

    override fun getHeaderRow(): List<String> {
        val headers =
            if (useV2FieldNames) {
                mutableListOf(
                    JavaBaseConstants.COLUMN_NAME_AB_RAW_ID,
                    JavaBaseConstants.COLUMN_NAME_AB_EXTRACTED_AT,
                    JavaBaseConstants.COLUMN_NAME_AB_META,
                    JavaBaseConstants.COLUMN_NAME_AB_GENERATION_ID,
                )
            } else {
                mutableListOf(
                    JavaBaseConstants.COLUMN_NAME_AB_ID,
                    JavaBaseConstants.COLUMN_NAME_EMITTED_AT,
                )
            }

        headers.addAll(recordHeaders)
        return headers
    }

    /** With root level flattening, the record columns are the first level fields of the json. */
    override fun getRecordColumns(json: JsonNode): List<String> {
        val values: MutableList<String> = LinkedList()
        for (field in recordHeaders) {
            val value = json[field]
            if (value == null || value.isNull) {
                values.add("")
            } else if (value.isValueNode) {
                // Call asText method on value nodes so that proper string
                // representation of json values can be returned by Jackson.
                // Otherwise, CSV printer will just call the toString method,
                // which can be problematic (e.g. text node will have extra
                // double quotation marks around its text value).
                values.add(value.asText())
            } else {
                values.add(Jsons.serialize(value))
            }
        }

        return values
    }
}
